### PR TITLE
feat: Go ast types for flux type expression syntax

### PR DESCRIPTION
**Pull request by jlapacik**
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: ast/ast.fbs [hunk 1]
```diff
@@ -1,5 +1,64 @@
...
+  kind:ParameterKind;
+}
+
+table FunctionType {
+  base_node:BaseNode;
+  parameters:[ParameterType];
+  retn:MonoType;
+}
+
+table Constraint {
```
Comment (nathanielc) [inline ast/ast.fbs:50]: Do we want to name this `TypeConstraint`? and maybe rename the Rust struct accordingly? Its a small thing but once we release it we can't easily change it.

File: ast/ast.go [hunk 2]
```diff
@@ -160,6 +160,93 @@ func (e Error) Error() string {
...
+	return "TypeExpression"
+}
+
+type MonoType interface {
+	Node
+	monotype()
+}
+
+type NamedType struct {
+	*Identifier
```
Comment (nathanielc) [inline ast/ast.go:179]: Why no BaseNode here?

File: ast/ast.fbs (hunk 1 above)
Comment (jlapacik) [reply ast/ast.fbs]: I agree.

**Comment by jlapacik** [comment]
@nathanielc would you like to give this a review? I've only finished the flatbuffer deserialization bits on the Go side, but it's probably much easier to review in pieces.

**Pull request merged**